server/plrhand.cpp

```diff
@@ -736,15 +741,24 @@ void player_update_last_war_action(struct player *pplayer)
 /**
    Handles a player cancelling a "pact" with another player.
 
-   packet.id is id of player we want to cancel a pact with
-   packet.val1 is a special value indicating what kind of treaty we want
-     to break. If this is CLAUSE_VISION we break shared vision. If it is
-     a pact treaty type, we break one pact level. If it is CLAUSE_LAST
-     we break _all_ treaties and go straight to war.
+   other_player_id is id of player we want to cancel a pact with
+   clause is a special value indicating what kind of treaty we want to
+     break. If this is CLAUSE_VISION we break shared vision. If it is a
+     pact treaty type, we break one pact level. If it is CLAUSE_LAST we
+     break _all_ treaties and go straight to war.
  */
 void handle_diplomacy_cancel_pact(struct player *pplayer,
                                   int other_player_id,
                                   enum clause_type clause)
+{
+  handle_diplomacy_cancel_pact_internal(pplayer, other_player_id, clause,
+                                        is_human(pplayer));
+}
+
+static void handle_diplomacy_cancel_pact_internal(struct player *pplayer,
```

Comment:
What's the reason for the new function? `protect_alliances` could just be deduced in the beginning

Reply:
We need the new function as once we start breaking alliances (at the `Check fall-out of a war declaration`), this function is called recursively with `protect_alliances` set to `false`.